### PR TITLE
known_hosts: support --diff

### PR DESCRIPTION
**Pull request by mgedmin**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
known_hosts

##### ANSIBLE VERSION

ansible 2.3.0

##### SUMMARY

This makes `known_hosts` return a diff when invoked with `--check --diff`.

### CODE REVIEW

**Comment by mgedmin** [comment]
I've tested this manually and it helped me figure out why ansible 2.3 reports `changed` for a couple of tasks in my playbook where ansible 2.2.1 didn't.

File: lib/ansible/modules/system/known_hosts.py
Comment (robinro) [inline lib/ansible/modules/system/known_hosts.py:122]: Imo `params` is not the right place for this. Since you once use the content here, why not call it `diff`?

File: lib/ansible/modules/system/known_hosts.py
Comment (robinro) [inline lib/ansible/modules/system/known_hosts.py]: That logic is tough to understand. I'd rather make it more linear (code untested):
```
old_lines = diff['before'].splitlines()
if replace_or_add or state=='absent':
    del old_lines[found_line]
diff[after] = '\n'.join(old_lines)
```

**Review by robinro — CHANGES_REQUESTED** [review]
Thanks @mgedmin for this PR. I'm looking forward to using this.

To ensure maintainability it's probably a good idea to cover this in the integration tests.
Add/remove a key in check_mode and ensure the diff looks as expected.

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:122]: This function returns `params`, which is then passed to `module.exit_json()`.  I need to set `params['diff']` to make the diff possible when `check_mode` is `False`.

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py]: Note: found_line is 1-based, so it should be `del old_lines[found_line - 1]`.

I'm reluctant to do refactorings without unit tests.  I'll see if I can manage to write one.

Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py]: The tests discovered one bug (the diff would show the key being added even when no changes would be done), so yay!

**Comment by mgedmin** [comment]
I've added an integration test (my first! not counting the small changes I've made to the authorized_diff integration test).

ready_for_review

File: test/integration/targets/known_hosts/aliases
Comment (mattclay) [inline test/integration/targets/known_hosts/aliases]: Does this integration test really require root?

Comment (mgedmin) [reply test/integration/targets/known_hosts/aliases]: I don't know!  I was copying what the authorized_key target did, without understanding what these things mean.

Comment (mattclay) [reply test/integration/targets/known_hosts/aliases]: You can run the test locally as a non-root user to see if it passes (make sure you remove the `needs/root` alias first):

`test/runner/ansible-test integration known_hosts -v`

Comment (mgedmin) [reply test/integration/targets/known_hosts/aliases]: I'm geting a weird failure if I do that:

```
TASK [known_hosts : add the same host in check mode] ***************************
task path: /home/mg/src/ansible/test/integration/targets/known_hosts/tasks/main.yml:63
fatal: [testhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "hash_host": false, 
            "key": "example.org ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAglyZmHHWskQ9wkh8LYbIqzvg99/oloneH7BaZ02ripJUy/2Zynv4tgUfm9fdXvAb1XXCEuTRnts9FBer87+voU0FPRgx3CfY9Sgr0FspUjnm4lqs53FIab1psddAaS7/F7lrnjl6VqBtPwMRQZG7qlml5uogGJwYJHxX0PGtsdoTJsM=\n", 
            "name": "example.org", 
            "path": "/home/mg/ansible_testing/known_hosts", 
            "state": "present"
        }, 
        "module_name": "known_hosts"
    }, 
    "msg": "ssh-keygen failed to hash host (rc=1,stdout='|1|cM06NS72gSewxKexCR+837GhoYU=|ht5OEhWMLmnhMLN2QJvKkSrlvjE= ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAglyZmHHWskQ9wkh8LYbIqzvg99/oloneH7BaZ02ripJUy/2Zynv4tgUfm9fdXvAb1XXCEuTRnts9FBer87+voU0FPRgx3CfY9Sgr0FspUjnm4lqs53FIab1psddAaS7/F7lrnjl6VqBtPwMRQZG7qlml5uogGJwYJHxX0PGtsdoTJsM=\n',stderr='')"
}
```

Which is, incidentally, the same failure I see on Shippable.

Hmm.

**Comment by mgedmin** [comment]
Fixed some Shippable failures, still have others.  E.g. `ssh-keygen` on CentOS 6 produces

```
line 5 invalid key: example.net...
/root/ansible_testing/known_hosts is not a valid known_hosts file.
```

because the old OpenSSH doesn't like `ssh-ed25519` keys, I think.

**Comment by mgedmin** [comment]
Rebased on current devel (partially because I messed up one of the refactoring commits and left a huge four-line one-liner of dead code).

**Comment by mgedmin** [comment]
The `dnf` failure on fedora25 is not my fault, honest!

ready_for_review

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [inline lib/ansible/modules/system/known_hosts.py:151]: many changes that don't have to do with diff

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [inline lib/ansible/modules/system/known_hosts.py:246]: why is 1 allowed? afaik this indicates an error in ssh-keygen

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:151]: The reason those changes are there:
- I was asked to write an integration test
- the integration test uncovered some minor buglets in the code
- I had to fix those buglets to make the new integration test pass

Incidentally, the extra buglets were these two:
- the known_hosts module didn't work under Python 3 (not a blocker, since modules aren't expected to work under Python 3, but I didn't like the yellow blob in Shippable and the fix wasn't large)
- the known_hosts module failed with an error if you specified `hash_host=true` and had a relatively recent OpenSSH version (6.4 or newer) -- looks like this was already filed as bug #20500.

Would you like me to extract the integration test and the bugfixes into a separate PR?  I would prefer not to (it's quite a bit of extra work for me), but I can be easily convinced with a little bit of blackmail (e.g. "we won't merge this unless you do that").

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:246]: Yes, this is a bug in ssh-keygen.  The commit message of 5eef669313594c809dfe66f27786618409ed91a0 has all the gory details.

Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:246]: BTW https://github.com/ansible/ansible/commit/dd9d5cc670eccf3c92c8bf974dd294787fe94169 was a workaround for a very similar issue.  And then https://github.com/ansible/ansible/commit/405c6820b0afd74fd335d2d302028e7e816f5bb9 added a 2nd invocation of ssh-keygen, but was apparently never tested with newer OpenSSH versions.

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [reply lib/ansible/modules/system/known_hosts.py:151]: no need for separate, just needed that explanation

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [reply lib/ansible/modules/system/known_hosts.py:246]: **sigh** i really hate compensating for these types of bugs .... but I don't see we have much of a choice. tempted to add a warning "your ssh is broken".

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [reply lib/ansible/modules/system/known_hosts.py:122]: i agree params is not the correct thing to use for return data .. but it would a separate fix for the module

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:246]: I feed kind of bad that I didn't report the bug upstream, but "OpenSSH" makes me think "maintained by OpenBSD guys", which leads to "probably in some CVS repository from 1998 with a mailing list instead of a bug tracker", and I don't have the energy.

(Also I'm probably wrong and they use Git, or at least Subversion now.  Still.)

File: lib/ansible/modules/system/known_hosts.py
Comment (mgedmin) [reply lib/ansible/modules/system/known_hosts.py:151]: I keep writing extensive commit messages and then forgetting reviewers have to take extra steps to see them in GitHub's review UI.

(And when I'm reviewing PRs for projects I maintain, I tend to ignore commit messages and just look at the final diff.)

Comment (bcoca) [reply lib/ansible/modules/system/known_hosts.py:151]: I normally rebase and put all messages in main commit (which shows in summary by default), but when you are doing several iterations its OK as is. When merging, we normally squash them if you hadn't already.

File: lib/ansible/modules/system/known_hosts.py
Comment (bcoca) [reply lib/ansible/modules/system/known_hosts.py:246]: FYI, https://www.openssh.com/report.html, they use bugzilla.

**Comment by mgedmin** [comment]
What's the status of this?  @bcoca's review indicates "requested changes", but it's not clear to me what changes are being requested.

**Review by bcoca — APPROVED** [review]

**Comment by mgedmin** [comment]
:sparkles: Thank you!  :sparkles:

**Pull request merged**
